builtin/reset.c

```diff
@@ -142,8 +144,41 @@ static void update_index_from_diff(struct diff_queue_struct *q,
 	for (i = 0; i < q->nr; i++) {
 		int pos;
 		struct diff_filespec *one = q->queue[i]->one;
+		struct diff_filespec *two = q->queue[i]->two;
 		int is_in_reset_tree = one->mode && !is_null_oid(&one->oid);
+		int is_missing = !(one->mode && !is_null_oid(&one->oid));
+		int was_missing = !two->mode && is_null_oid(&two->oid);
 		struct cache_entry *ce;
+		struct cache_entry *ceBefore;
+		struct checkout state = CHECKOUT_INIT;
+
+		/*
+		 * When using a virtual filesystem, the cache entries that are
+		 * added here will not have the skip-worktree bit set.
+		 * Without this code there is data that is lost because the files that
+		 * would normally be in the working directory are not there and show as
+		 * deleted for the next status or in the case of added files just disappear.
+		 * We need to create the previous version of the files in the working
+		 * directory so that they will have the right content and the next
+		 * status call will show modified or untracked files correctly.
```

Comment:
I'm kind of confused - why is the VFS case different from the sparse-checkout case? That is, if a file had `SKIP_WORKTREE` enabled before, shouldn't the solution be to apply it after the `reset`, rather than checking out the file altogether? Theoretically, the code below does that, so I'm not sure what's causing the file to appear deleted after `reset`.

I also don't know how this bit interacts with the fix that went upstream - wouldn't this checkout the file, then immediately mark it as `SKIP_WORKTREE`?

I remember that the "information lost" reason was the original logic behind this implementation, but it doesn't square with what `SKIP_WORKTREE` means (i.e., I don't care what's in the worktree). Is `SKIP_WORKTREE` used differently in VFS for Git?

Reply:
> I'm kind of confused - why is the VFS case different from the sparse-checkout case? That is, if a file had `SKIP_WORKTREE` enabled before, shouldn't the solution be to apply it after the `reset`, rather than checking out the file altogether? Theoretically, the code below does that, so I'm not sure what's causing the file to appear deleted after `reset`.

In VFS for Git, `SKIP_WORKTREE` really means "this is a virtual file". This means we need to be careful about how we interact with the filesystem. In particular, any change to the worktree that Git makes will trigger the virtual filesystem in ways that are important.

The `SKIP_WORKTREE` bits are also dictated by the virtual filesystem layer. There isn't actually a set of patterns to check. Once the file is written, the virtualization layer says "this is a file we care about" and the `SKIP_WORKTREE` bit is removed at that layer.

> I also don't know how this bit interacts with the fix that went upstream - wouldn't this checkout the file, then immediately mark it as `SKIP_WORKTREE`?

No, I don't think it does. If `core_virtualfilesystem` is enabled, then `path_in_sparse_checkout()` is always true. See...

```c
static int path_in_sparse_checkout_1(const char *path,
				     struct index_state *istate,
				     int require_cone_mode)
{
	int dtype = DT_REG;
	enum pattern_match_result match = UNDECIDED;
	const char *end, *slash;

	/*
	 * When using a virtual filesystem, there aren't really patterns
	 * to follow, but be extra careful to skip this check.
	 */
	if (core_virtualfilesystem)
		return 1;
```

> I remember that the "information lost" reason was the original logic behind this implementation, but it doesn't square with what `SKIP_WORKTREE` means (i.e., I don't care what's in the worktree). Is `SKIP_WORKTREE` used differently in VFS for Git?

It does mean that. In VFS for Git, there isn't any idea of "I have chosen that these files aren't important" but instead "I have not loaded these files in the filesystem yet". With a fresh repo, every file is virtual (even an `ls` only doesan object size query, not a full population of the files).

The test that used to exist as `t7114` in 570f64b7225ff48a0061a5dab4233524cdb2f538 shows the behavior that is incorrect from the sparse-checkout standpoint.

```sh
test_expect_success 'setup' '
        test_tick &&
        echo "checkout file" >c &&
        echo "modify file" >m &&
        echo "delete file" >d &&
        git add . &&
        git commit -m "initial commit" &&
        echo "added file" >a &&
        echo "modification of a file" >m &&
        git rm d &&
        git add . &&
        git commit -m "second commit" &&
        git checkout -b endCommit
'

test_expect_success 'reset when there is a sparse-checkout' '
        echo "/c" >.git/info/sparse-checkout &&
        test_config core.sparsecheckout true &&
        git checkout -B resetBranch &&
        test_path_is_missing m &&
        test_path_is_missing a &&
        test_path_is_missing d &&
        git reset HEAD~1 &&
        test "checkout file" = "$(cat c)" &&
        test "modification of a file" = "$(cat m)" &&
        test "added file" = "$(cat a)" &&
        test_path_is_missing d
'
```

Look at this test: we do a `git reset HEAD~1` and we expect that to populate `m` and `a`, despite them being outside of the sparse-checkout patterns. The `SKIP_WORKTREE` bit means (in sparse-checkout world) "Git doesn't care about these files on-disk, so it's ok to lose the data". In VFS for Git, we need to populate the files so they become important to the virtual filesystem (which drops the `SKIP_WORKTREE` bit).

Thanks for asking the deep questions, @vdye. Hopefully this helps clarify the situation a bit more?

Reply:
I think the part that I'm missing is that the function in its current state should preserve `SKIP_WORKTREE` regardless of whether `sparse-checkout` is in use. Specifically, this condition:

```
if((pos >= 0 && ce_skip_worktree(active_cache[pos])) ||
   ...)
        ce->ce_flags |= CE_SKIP_WORKTREE;
```

If the file exists in the index, and it already has `SKIP_WORKTREE` enabled, then the resulting file will also have `SKIP_WORKTREE` enabled.

The code block you added only does something if `SKIP_WORKTREE` is already enabled on the original entry, so I'm confused as to why the code block I quoted isn't applying `SKIP_WORKTREE` to the new entry.

I also don't see why this is true:
> In VFS for Git, we need to populate the files so they become important to the virtual filesystem (which drops the SKIP_WORKTREE bit).

If the file had `SKIP_WORKTREE` before the reset, why do we care about it after the reset?

Sorry if I'm missing something - I remember this solution upstream being general to _any_ use of `SKIP_WORKTREE`, so I want to understand why the VFS for Git usage of it differs from all others.

Reply:
> I think the part that I'm missing is that the function in its current state should preserve `SKIP_WORKTREE` regardless of whether `sparse-checkout` is in use.

I think this is really an issue with the "muddiness" of the virtual filesystem, where the `SKIP_WORKTREE` bit gets cleared later due to the virtual filesystem hook.

But this also made me look closer at the implementation and find that the _reason_ this works in the virtual filesystem is that `file_exists()` check that actually populates the file. The rest of the block is not necessary. I have pushed a new commit and will update the PR description shortly.